### PR TITLE
Use less memory for stateful material storage

### PR DESCRIPTION
**Pull request by permcody**
I'm dropping TBB concurrent map because it seems to use way, way more memory than it should. We are going to just use std::unordered_map until somebody comes up with something better.

### CODE REVIEW

File: framework/include/utils/HashMap.h [hunk 1]
```diff
@@ -15,51 +15,14 @@
...
-};
-
-#else
-
-// Use a hash table
 #include <unordered_map>
 
-template<typename Key, typename T> /*, typename Hash = std::tr1::hash<Key>, class Pred = std::equal_to<Key>, typename Allocator = std::allocator<std::pair<const Key,T> > >*/
+template<typename Key, typename T>
 class HashMap : public std::unordered_map<Key, T> /*, Hash, Pred, Allocator >*/
```
Comment (andrsd) [inline framework/include/utils/HashMap.h]: Those `/*, Hash, Pred, Allocator >*/` comments can probably go away as well...

File: test/src/materials/StatefulTest.C [hunk 2]
```diff
@@ -17,32 +17,53 @@ template<>
...
-    _coupled_val(isParamValid("coupled") ? &coupledNodalValue("coupled") : nullptr)
+    _coupled_val(isParamValid("coupled") ? &coupledNodalValue("coupled") : nullptr),
+    _prop_names(getParam<std::vector<std::string> >("prop_names")),
+    _prop_values(getParam<std::vector<Real> >("prop_values"))
 {
+  unsigned int num_names = _prop_names.size();
+  unsigned int num_values = _prop_values.size();
+
+  if (num_names != num_values)
+    mooseError("Number of prop_names must match the number of prop_values for a GenericConstantMaterial!");
```
Comment (andrsd) [inline test/src/materials/StatefulTest.C]: How is this related to `GenericConstantMaterial`?

File: test/tests/materials/stateful_prop/many_stateful_props.i [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+    type = ElementAverageValue
+    variable = prop1
+    execute_on = 'initial timestep_end'
+  [../]
+[]
+
+[Executioner]
+  type = Transient
+
+  # Preconditioned JFNK (default)
```
Comment (andrsd) [inline test/tests/materials/stateful_prop/many_stateful_props.i]: We need to stop copying this comment around...

File: test/tests/materials/stateful_prop/many_stateful_props.i [hunk 4]
```diff
@@ -0,0 +1,129 @@
+# This test creates several unused stateful properties.
+# It's here to make sure that we don't consume too much
+# memory if we store them all. With 180x180 elements
+# we were previously seeing nearly a Gigabyte of memory
+# consumed using TBB's map. We are now using unordered
+# map which saves us 6x to 8x on memory.
```
Comment (andrsd) [inline test/tests/materials/stateful_prop/many_stateful_props.i]: Maybe a better place for this performance discussion would be the actual commit message that changed from TBB map to unordered map.

**Review by andrsd — CHANGES_REQUESTED** [review]

**Comment by rwcarlsen** [comment]
I'd prefer that we just manually change the use of ``HashMap`` to ``std::unordered_map`` for the stateful property handling rather than a typedef which just adds indirection.

**Comment by andrsd** [comment]
True, `HashMap` can actually go away. It was created to have different thread-safe hash maps for different threading models.

**Comment by permcody** [comment]
Well it does serve one important purpose. It currently has the thread lock in it. If we remove this class we'll have to carefully find all threaded access to the map and put the lock in each of those locations. I think we may only have three (ComputeMaterialsThread, prolongStatefulProps, restrictStatefulProps) but it'll take some time to go through and verify all of that. Plus it'll be error prone as people work on the code and accidentally add insertions in threaded regions. 

Still, it may be the right solution but not for this PR. I'll address the other comments and we'll get this in as-is #8487

File: test/tests/materials/stateful_prop/many_stateful_props.i (hunk 4 above)
Comment (permcody) [inline test/tests/materials/stateful_prop/many_stateful_props.i]: Maybe, However I think our test files should be much more descriptive of what they are testing. Users/Developers should be able to access this information without diving into commit messages.

**Review by andrsd — APPROVED** [review]

**Pull request merged**